pkg/client/clientcmd/loader.go

```diff
@@ -33,31 +33,35 @@ import (
 const (
 	RecommendedConfigPathFlag   = "kubeconfig"
 	RecommendedConfigPathEnvVar = "KUBECONFIG"
+
+	DefaultEnvVarIndex     = 0
+	DefaultCurrentDirIndex = 1
+	DefaultHomeDirIndex    = 2
 )
 
-// ClientConfigLoadingRules is a struct that calls our specific locations that are used for merging together a Config
+// ClientConfigLoadingRules is an ExplicitPath and string slice of specific locations that are used for merging together a Config
+// Callers can put the chain together however they want, but we'd recommend:
+// [0] = EnvVarPath
+// [1] = CurrentDirectoryPath
+// [2] = HomeDirectoryPath
+// ExplicitPath is special, because if a user specifically requests a certain file be used and error is reported if thie file is not present
 type ClientConfigLoadingRules struct {
-	CommandLinePath      string
-	EnvVarPath           string
-	CurrentDirectoryPath string
-	HomeDirectoryPath    string
+	ExplicitPath string
+	Precedence   []string
 }
 
-// NewClientConfigLoadingRules returns a ClientConfigLoadingRules object with default fields filled in.  You are not required to
+// NewDefaultClientConfigLoadingRules returns a ClientConfigLoadingRules object with default fields filled in.  You are not required to
 // use this constructor
-func NewClientConfigLoadingRules() *ClientConfigLoadingRules {
+func NewDefaultClientConfigLoadingRules() *ClientConfigLoadingRules {
```

Comment:
I'd like a method that takes the precedence as argument, say `NewClientConfigLoadingRules(precedence []string)`, so if I'm creating a completely custom set of precedence rules I don't get the default ones created before setting `Precedence`.